crosslink/internal/tidy.go

```diff
@@ -0,0 +1,204 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crosslink
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	"go.uber.org/zap"
+	"golang.org/x/mod/modfile"
+)
+
+type graphNode struct {
+	file    *modfile.File
+	path    string
+	name    string
+	deps    []string
+	index   int
+	sccRoot int
+	onStack bool
+}
+
+func Tidy(rc RunConfig, outputPath string) error {
+	rc.Logger.Debug("crosslink run config", zap.Any("run_config", rc))
+
+	rootModule, err := identifyRootModule(rc.RootPath)
+	if err != nil {
+		return fmt.Errorf("failed to identify root module: %w", err)
+	}
+
+	// Read circular dependency allowlist
+
+	var allowCircular []string
+	if rc.AllowCircular != "" {
+		var file *os.File
+		file, err = os.Open(rc.AllowCircular)
+		if err != nil {
+			return fmt.Errorf("failed to open cicular dependency allowlist: %w", err)
+		}
+		defer file.Close()
+		scanner := bufio.NewScanner(file)
+		for scanner.Scan() {
+			line := scanner.Text()
+			if line != "" {
+				allowCircular = append(allowCircular, line)
+			}
+		}
+		if err = scanner.Err(); err != nil {
+			return fmt.Errorf("failed to read cicular dependency allowlist: %w", err)
+		}
+	}
+
+	// Read intra-repository dependency graph
+
+	graph := make(map[string]*graphNode)
+	var modsAlpha []string
+	err = forGoModFiles(rc, func(filePath string, name string, file *modfile.File) error {
+		if !strings.HasPrefix(name, rootModule) {
+			rc.Logger.Debug("ignoring module outside root module namespace", zap.String("mod_name", name))
+			return nil
+		}
+		modsAlpha = append(modsAlpha, name)
+		modPath, _ := strings.CutSuffix(filePath, "/go.mod")
```

Comment:
Why do we need to do this?

Reply:
The `strings.CutSuffix` call? The `filePath` contains the relative path to the `go.mod` file of the module, so we remove `/go.mod` to get the path to the module directory.

Reply:
Should we check the `found` return argument then and fail if it is `false`?

Reply:
I feel like it is useful in case the surrounding code changes to be able to more easily debug this (if we ever stop filtering for some reason, we will quickly know that this is failing)